README
SYNOPSIS

    In dist.ini:

     [Rinci::Validate]

    In your module:

     $SPEC{foo} = {
         args => {
             arg1 => { schema => ['int*', default=>3] },
             arg2 => { },
         },
     };
     sub foo {
         my %args = @_;
    
         my $arg1 = $args{arg1}; # VALIDATE_ARG
         ...
     }

    output will be something like:

     $SPEC{foo} = {
         args => {
             arg1 => { schema => ['int*', default=>3] },
             arg2 => { },
         },
     };
     sub foo {
         my %args = @_;
    
         my $arg1 = $args{arg1}; require Scalar::Util::Numeric; my $arg_err; (($arg1 //= 3), 1) && ((defined($arg1)) ? 1 : (($err_arg1 = 'TMPERRMSG: required data not specified'),0)) && ((Scalar::Util::Numeric::isint($arg1)) ? 1 : (($err_arg1 = 'TMPERRMSG: type check failed'),0)); return [400, "Invalid value for arg1: $err_arg1"] if $arg1; # VALIDATE_ARG
         ...
     }

    You can also validate all arguments:

     sub foo {
         my %args = @_; # VALIDATE_ARGS
    
         ...
     }

DESCRIPTION

    This plugin inserts argument validation code into your module source
    code, at location marked with # VALIDATE_ARG or # VALIDATE_ARGS.
    Validation code is compiled using Data::Sah from Sah schemas specified
    in args property in Rinci function metadata in the module.

    This plugin detects the more recently developed plugin
    Dist::Zilla::Plugin::Rinci::Wrap and will skip running if the latter is
    loaded.

 USAGE

    To validate a single argument, in your module:

     sub foo {
         my %args = @_;
         my $arg1 = $args{arg1}; # VALIDATE_ARG

    The significant part that is interpreted by this module is my $arg1.
    Argument name is taken from the lexical variable's name (in this case,
    arg1). Argument must be defined in the args property of the function
    metadata. If argument name is different from lexical variable name,
    then you need to say:

     my $f = $args->{frobnicate}; # VALIDATE_ARG frobnicate

    To validate all arguments of the subroutine, you can say:

     sub foo {
         my %args = @_; # VALIDATE_ARGS

    There should only be one VALIDATE_ARGS per subroutine.

    If you use this plugin, and you plan to wrap your functions too using
    Perinci::Sub::Wrapper (or through Perinci::Access, Perinci::CmdLine,
    etc), you might also want to put
    x.perinci.sub.wrapper.disable_validate_args => 1 attribute into your
    function metadata, to instruct Perinci::Sub::Wrapper to skip generating
    argument validation code when your function is wrapped, as argument
    validation is already done by the generated code.

    If there is an unvalidated argument, this plugin will emit a warning
    notice. To skip validating an argument (silence the warning), you can
    use:

     sub foo {
         my %args = @_;
         my $arg1 = $args{arg1}; # NO_VALIDATE_ARG

    or:

     sub foo {
         # NO_VALIDATE_ARGS

FAQ

 Rationale for this plugin?

    This plugin is an alternative to Perinci::Sub::Wrapper, at least when
    it comes to validating arguments. Perinci::Sub::Wrapper can also
    generate argument validation code (among other things), but it is done
    during runtime and can add to startup overhead (compiling complex
    schemas for several subroutines can take up to 100ms or more, on my
    laptop). Using this plugin, argument validation code is generated
    during building of your distribution.

    Using this plugin also makes sure that argument is validated whether
    your subroutine is wrapped or not. Using this plugin also avoids
    wrapping and adding nest level, if that is not to your liking.

    Instead of using this plugin, you can use wrapping either by using
    Perinci::Exporter or by calling Perinci::Sub::Wrapper's wrap_sub
    directly.

 But why use Rinci metadata or Sah schema?

    In short, adding Rinci metadata to your subroutines allows various
    tools to do useful stuffs, relieving you from doing those stuffs
    manually. Using Sah schema allows you to write validation code
    succintly, and gives you the ability to automatically generate
    Perl/JavaScript/error messages from the schema.

    See their respective documentation for more details.

 But the generated code looks ugly!

    Admittedly, yes. Validation source code is formatted as a single long
    line to avoid modifying line numbers, which is desirable when debugging
    your modules. An option to not compress everything as a single line
    might be added in the future.



Changes
Revision history for Dist-Zilla-Rinci-Validate

0.20    2015-01-03 (PERLANCAR)

	- No functional changes.

	- Rebuild (Generate TODO.md).


0.19    2014-10-29 (PERLANCAR)

	- [Bugfix] Schema default and default from 'default' argument spec
	  property were not applied.


0.18    2014-09-12 (PERLANCAR)

	- No functional changes.

	- Fix syntax error (d'oh).


0.17    2014-09-12 (PERLANCAR)

	- No functional changes.

	- Tweak die message: add sub name prefix to make it clearer.


0.16    2014-09-12 (PERLANCAR)

	- Bugfix: allow args_as=array/whatever if we're validating arg one
	  by one.

	- Bugfix: Avoid use of Perinci::Access::Perl to avoid changing
	  result_naked, args_as.


0.15    2014-09-06 (PERLANCAR)

	- Adjust to Perinci::Sub::Wrapper 0.62
	  (_perinci.sub.wrapper.validate_args is now renamed).


0.14     2014-06-27 (SHARYANTO)

         - Fix regex for package declaration so false positives are reduced.


0.13     2014-04-30 (SHARYANTO)

         - Adjust to Perinci::Access::Perl 0.66 (no longer has
           extra_wrapper_args, but remove_internal_properties is now off by
           default in Perinci::Sub::Normalize).


0.12     2014-03-08 (SHARYANTO)

         - Skip running if DZP::Rinci::Wrap is loaded.


0.11     2013-12-18 (SHARYANTO)

         - No functional changes. Fix coding error.


0.10     2013-12-18 (SHARYANTO)

         - No functional changes. Adjust to Perinci::Access::Perl 0.54+.


0.09     2013-09-11 (SHARYANTO)

         - No functional changes. Adjust to Data::Sah 0.18.


0.08     2013-09-11 (SHARYANTO)

         - No functional changes. Add 'use experimental smartmatch' for Perl
           5.18+".


0.07     2013-09-08 (SHARYANTO)

         - Bugfix: make sure all uri's fed to Perinci::Access::Perl are pl: uris.


0.06     2013-09-08 (SHARYANTO)

         - No functional changes. Replace Perinci::Access::InProcess with
           ::Perl.


0.05     2013-03-02 (SHARYANTO)

         [BUG FIXES]

         - Validate (to set default value) non-required arguments if their
           schemas specify default values.


0.04     2013-03-02 (SHARYANTO)

         [BUG FIXES]

         - Don't validate non-required (req=>0) arguments if they are not
           passed.


0.03     2013-01-11 (SHARYANTO)

         - No functional changes.

         [BUG FIXES]

         - Make sure data_name is alphanum only (foo.contains -> foo_contains).

         - Wrapping fixes (declare variables and require modules required by the
           generated validator code).


0.02     2012-11-06 (SHARYANTO)

         - Add # NO_VALIDATE_ARG(S?)

         - Add notice about setting _perinci.sub.wrapper.validate_args => 0 if
           validated arguments.

         - Some fixes.


0.01     2012-11-06 (SHARYANTO)

         - First release.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Insert argument validator code in output code",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-Rinci-Validate",
  "dist_version" => "0.07",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Plugin::Rinci::Validate",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Data::Sah" => "0.10",
    "Dist::Zilla" => 0,
    "Log::Any" => 0,
    "Perinci::Access::Perl" => "0.38"
  },
  "script_files" => [],
  "test_requires" => {
    "Capture::Tiny" => 0,
    "Test::More" => "0.88",
    "blib" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Insert argument validator code in output code",
  "AUTHOR" => "perlancar <perlancar\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Perl::osnames" => "0.09"
  },
  "DISTNAME" => "Dist-Zilla-Plugin-Rinci-Validate",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Dist::Zilla::Plugin::Rinci::Validate",
  "PREREQ_PM" => {
    "Data::Sah" => "0.18",
    "Dist::Zilla" => 0,
    "Perinci::Sub::Normalize" => 0,
    "experimental" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.20",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Sah" => "0.18",
  "Dist::Zilla" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Perinci::Sub::Normalize" => 0,
  "Perl::osnames" => "0.09",
  "Test::More" => 0,
  "experimental" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



